src/odemis/acq/millmng.py

```diff
@@ -0,0 +1,533 @@
+            future._task_state = CANCELLED
+            future.running_subf.cancel()
+            logging.debug("Milling procedure cancelled.")  # add line number
+        return True
+
+    def estimate_milling_time(self, site_idx: int = 0, actual_time_per_site: float = None) -> float:
+        """
+        Estimates the milling time for the given feature.
+        :param site_idx: (int) index of the site to be milled
+        :param actual_time_per_site: (float) actual milling time measured for a single site
+        :return: (float > 0): the estimated time
+        """
+        if len(self.sites) == 0:
+            logging.debug("Site location is not provided by the user, cancelling the milling")
+            self._future._task_state = CANCELLED
+            raise CancelledError()
+
+        remaining_sites = (len(self.sites) - site_idx)
+
+        if actual_time_per_site:
+            milling_time = actual_time_per_site * remaining_sites
+        else:
+            milling_time = self._milling_time_estimate * remaining_sites
+
+        self._remaining_stage_time -= self.estimate_stage_movement_time(self.sites[site_idx])
+        total_duration = milling_time + self._remaining_stage_time
+
+        return total_duration
+
+    def estimate_stage_movement_time(self, site: CryoFeature, stage_pos_ref=None) -> float:
+        """
+        Estimation the time taken by the stage to move from current position to the location of given site
+        :param site: (CryoFeature) consists X,Y,Z coordinates for stage location in m.
+        :return: (float) time to move the stage between two points in s.
+        """
+        # current position from x,y,z of stage position and eliminating rx,ry,rz
+        if stage_pos_ref is None:
+            current_pos = [p[1] for p in self._stage.position.value.items() if len(p[0]) == 1]
+        else:
+            current_pos = stage_pos_ref
+        target_pos = site.pos.value  # list
+        diff = [abs(target - current) for target, current in zip(target_pos, current_pos)]
+        stage_time = math.sqrt(sum(d ** 2 for d in diff)) / self._move_speed
+
+        return stage_time
+
+    def estimate_drift_time(self, setting: MillingSettings) -> float:
+        """
+        Estimate the time taken to acquire drift correction images
+        :param setting: (MillingSettings) milling settings
+        return (float): estimated time to acquire 1 anchor area
+        """
+        if setting.dcRoi.value == UNDEFINED_ROI:
+            return 0
+
+        nb_anchor_scanning = math.ceil(setting.duration.value / setting.dcPeriod.value)
+        drift_estimation = AnchoredEstimator(self._scanner, self._detector,
+                                             setting.dcRoi.value, setting.dcDwellTime.value, ANCHOR_MAX_PIXELS,
+                                             follow_drift=False)
+        return drift_estimation.estimateAcquisitionTime() * nb_anchor_scanning
+
+    def _restore_beam_shift(self, current: float):
```

Comment:
I added this function which changes current and restores the previous shift. The previous shift should already be drift compensated due to which I do not separately compensate for the drift. This approach is a little bit different from your suggestion to add drift. Is this way good to go?